### ISSUE TITLE
[Proposal]: 'constructor' Keyword as Constructor Name

### ISSUE THREAD

**Issue by AJLProjects** [dotnet/csharplang#4267]
# 'constructor' Keyword as Constructor Name

* [x] Proposed
* [ ] Prototype: Not Started
* [ ] Implementation: Not Started
* [ ] Specification: Not Started

## Summary
[summary]: Currently when you create a constructor, the name must match the class name. 
This is a form of redundancy since the name can be contextually implied from the class, so my proposal is to allow 'constructor' to be used as a constructor name. This is also useful for maintainablity since when you are changing the name of the class, the constructor name will then not need to be changed. The current approach violates the DRY principle.

## Motivation
[motivation]: 
- Reducing redundancy
- Visual simplicity
- Slightly more maintainable code

```
class SomeClass // <<<<< EXACTLY THE SAME
{
  public SomeClass() // <<<<<< EXACTLY THE SAME
  {

  }
}
```

## Detailed design
[design]: 
Simply put, allow 'constructor' keyword to be used as a constructor name.  
The rest works exaclty the same e.g. constructor overloading.

```
class A
{
   private int value;
   constructor(int value)
  {
    this.value = value;
  }

   constructor(int valueA, int valueB)
  {
    this.value = valueA + valueB;
  }
}
```
## Drawbacks
[drawbacks]: 
- Having competing conventions with the current approach (But then var is a competing convention as well)

## Alternatives
[alternatives]: 
- Use 'Constructor' with a capital C 

## Unresolved questions
[unresolved]: 
N/A

## Design meetings
None, just a suggestion.

**Comment by YairHalberstadt** [comment]
Thanks for your suggestion!

This has been discussed before at https://github.com/dotnet/csharplang/discussions/1928

The consensus seemed to be the advantage doesn't outweigh the cost of having two syntaxes for the same thing. To quote an LDC member:

> It seems highly unlikely C# would change here. Having two ways to do something, where the new way provides only a tiny amount of marginal benefit, would likely not be desirable.

As such I'm going to close this so that discussion can continue over there all in one place. I don't think it's worth opening a new issue unless there's a new reason/use case which is significantly more compelling, and it seems likely a formal proposal would be accepted.
